### DISCUSSION TITLE
Allow searching for .c4 in node_modules: Where to start?

### DISCUSSION THREAD

**Discussion by rumpelrausch** [likec4/likec4#1174] [Q&A] [answered]
For a current project I'd like to include the node_modules folder when searching for .c4 sources.
I found that the VSCode plugin _does_ search inside node_modules, but the likec4 cli doesn't. I suppose it's some exclude configuration of vite.

I could either support with a PR or just build my own fork, but I'm rather lost with the overall complexity.
So:
Could anyone tell me where to start?

**Comment by davydkov** [comment, marked as answer]
Hey @rumpelrausch 

I think [this check in langium](https://github.com/eclipse-langium/langium/blob/516fe4cc0a520d9a88e53409f5effd1aa3b6f2eb/packages/langium/src/workspace/workspace-manager.ts#L186) prevents recursive search in `node_modules`.

You can override this method in [LikeC4WorkspaceManager](https://github.com/likec4/likec4/blob/main/packages/language-server/src/shared/WorkspaceManager.ts#L9) if you need it asap.

I think this should be configurable, like `includes/excludes` in tsconfig, but this requires more effort

**Reply by rumpelrausch** [reply]
Ah, thanks, and now I see why the VSCode extension works differently:
It implements its own workspace logics which are purely based upon glob.

IMHO VSCode and CLI should behave identically. I'd expect this workflow:
1. Develop a C4 document interactively using VSCode.
2. Store the results as source code (allowing peer reviews).
3. Build a static result manually or via some build pipeline.

The results must be the same, even if this would mean excluding `node_modules` from the VSCode plugin globber.

---

As far as I can see, langium excludes `node_modules` because it follows a general approach: With whatever DSL a project might define, some sources may or may not be located inside `node_modules` but aren't part of the current package code. Fair enough.

So, overwriting `DefaultWorkspaceManager.includeEntry` is a viable option, and it should be configurable via CLI parameters.

---

In the meantime, may I suggest using the same globbing in VSCode as in langium (excluding dot files and `node_modules`)?

**Reply by davydkov** [reply]
> may I suggest using the same globbing in VSCode as in langium (excluding dot files and node_modules)?

Reasonable suggestion, agree.
I remember there was a thread on same topic in langium discussions. I check it again
